storage/include/sparse_merkle/internal_node.h

```diff
@@ -0,0 +1,283 @@
+  //
+  // Return the index of the bottom most InternalNode created.
+  size_t insertInternalChildren(size_t index, Version version, Nibble child_key, size_t bits_in_common);
+
+  // After the collission and the creation of new InternalChild nodes, there is
+  // still room in this BatchedInternalNode to add the previously stored
+  // LeafChild and the new LeafChild. Go ahead and do that insertion in this
+  // function.
+  //
+  // Always return InsertComplete{}.
+  InsertResult insertTwoLeafChildren(
+      size_t index, Version version, Nibble child_key, size_t bits_in_common, LeafChild child1, LeafChild child2);
+
+  // Return true if the index does not contain a child.
+  bool isEmpty(size_t index) const { return !children_.at(index).has_value(); }
+
+  // Return true if the index has a child that contains an internal node.
+  bool isInternal(size_t index) const;
+
+  // Attempt to insert into an empty child.
+  //
+  // Return InsertComplete if the node was empty.
+  // Otherwise return std::nullopt.
+  std::optional<InsertResult> insertIntoEmptyChild(size_t index, const LeafChild& child);
+
+  // Attempt to overwrite a matching key at index.
+  //
+  // Return InsertComplete if the key was overwritten.
+  // Otherwise return std::nullopt.
+  std::optional<InsertResult> overwrite(size_t index, const LeafChild& child);
+
+  // There has been a collision of leaf keys that do not match.
+  //
+  // One or more parents must be created to store them.
+  //
+  // `depth` is the current depth of this BatchedInternalNode in ths sparse merkle
+  // tree.
+  //
+  // Return either CreateNewBatchedInternalNodes or InsertComplete.
+  InsertResult splitNode(size_t index, const LeafChild& child, size_t depth);
+
+  // Return the hash of a given index.
+  const Hash& getHash(size_t index) const;
+
+  // Walk the tree from `index` up to the root of this BatchedInternalNode to
+  // compute the root hash.
+  //
+  // Update the hashes and versions of nodes along the way.
+  void updateHashes(size_t index, Version version);
+
+  // Return the the height of the node at the given index.
+  //
+  // The height depends upon MAX_CHILDREN, which dictates the maximum depth of a
+  // binary tree.
+  size_t height(size_t index) const;
+
+  // Return the index of the parent node, given the index of the child node.
+  //
+  // Return std::nullopt if the index points to the root of this
+  // BatchedInternalNode.
+  std::optional<size_t> parentIndex(size_t index) {
+    if (index == 0) return std::nullopt;
+    return (index - 1) / 2;
+  }
+
+  // Return the index of the left child of a node at a given index.
+  size_t leftChildIndex(size_t index) const { return 2 * index + 1; }
+
+  // Return the index of the right child of a node at a given index.
+  size_t rightChildIndex(size_t index) const { return 2 * index + 2; }
+
+  // Is this node a left child?
+  bool isLeftChild(size_t index) const { return index % 2 != 0; }
```

Comment:
[Minor]: Does it make sense to move these to the cpp file with internal linkage as they are private and don't use member variables?

Reply:
Hmmm. Will consider it for a future commit.